### PR TITLE
CDAP-3435 Program status shows Running even after it is stopped.

### PR DESCRIPTION
**Pull request by henry-cask**
When doing fix for CDAP-3374 we added access to transaction within transaction to reduce logging messages which cause problems.

This PR removes the nested the call to transactions inside a transaction (call to retrieveProgramIdForRunRecord actually cause call to dataStore which has transaction) for the running run records fixer thread.
Add additional log to debug run records cleaning. Cache the program ids for run ids when filtering for valid program type.

### CODE REVIEW

**Comment by henry-cask** [comment]
Build: http://builds.cask.co/browse/CDAP-DUT2695-2

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ProgramLifecycleService.java [hunk 1]
```diff
@@ -257,35 +259,63 @@ private void validateAndCorrectRunningRunRecords() {
...
+    LOG.debug("End getting {} run records not actually running.", notActuallyRunning.size());
+
+    final Map<String, Id.Program> runIdToProgramId = new HashMap<>();
+
+    LOG.debug("Start getting invalid run records  ...");
+    Collection<RunRecordMeta> invalidRunRecords =
+      Collections2.filter(notActuallyRunning, new Predicate<RunRecordMeta>() {
+        @Override
+        public boolean apply(@Nullable RunRecordMeta input) {
+          if (input == null) {
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ProgramLifecycleService.java]: can input be null?

Comment (henry-cask) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ProgramLifecycleService.java]: Should not be. I will take out the null check for both filters.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ProgramLifecycleService.java [hunk 2]
```diff
@@ -468,6 +492,8 @@ public void run() {
         RunRecordsCorrectorRunnable.LOG.debug("End correcting invalid run records.");
       } catch (Throwable t) {
         // Ignore any exception thrown since this behaves like daemon thread.
+        LOG.warn("Unable to complete correcting run records: {}",
+                 (t.getCause() != null ? t.getCause().getMessage() : t.getMessage()));
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ProgramLifecycleService.java]: Use `Throwables.getRootCause(t)`

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ProgramLifecycleService.java [hunk 3]
```diff
@@ -257,35 +259,57 @@ private void validateAndCorrectRunningRunRecords() {
...
-        return (targetProgramId != null) && !runIdToRuntimeInfo.containsKey(RunIds.fromString(runId));
+        // Check if it is not actually running.
+        return !runIdToRuntimeInfo.containsKey(RunIds.fromString(runId));
       }
     });
+    LOG.debug("End getting {} run records not actually running.", notActuallyRunning.size());
+
+    final Map<String, Id.Program> runIdToProgramId = new HashMap<>();
+
+    LOG.debug("Start getting invalid run records  ...");
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ProgramLifecycleService.java]: More like a trace to me.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ProgramLifecycleService.java (hunk 2 above)
Comment (henry-cask) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ProgramLifecycleService.java]: Sounds good

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ProgramLifecycleService.java (hunk 3 above)
Comment (henry-cask) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ProgramLifecycleService.java]: Will make the change.

**Comment by henry-cask** [comment]
Last build passes: http://builds.cask.co/browse/CDAP-DUT2695-5

**Pull request merged**